rstudio/rstudio
# New feature: customize project display name

### Intent

Addresses #1909 and #14029

### Approach

Users can specify a custom name for an R project in the Project Options dialog. This displays instead of the name of the project folder in the window title for Desktop, the browser tab for Server, and the dock icon on macOS. The custom name is appended (in parentheses) to entries in the File / Recent Projects menu and the Projects dropdown. The project dropdown button itself displays the custom name instead of the folder name (but the tooltip for that button still shows the path including the .Rproj file).

The project name is stored in the `.Rproj` file. If you set a name and then open the project in an older version of RStudio and it rewrites the project file, the name will be lost. This is true for anything we add to this file; a given RStudio version will only write back the fields it knows about and fields added by a newer RStudio will be lost.

The project MRU data is being stored in a new file (`project_name_mru`) where each line starts with the path to the `.Rproj` (exactly the same as in the old `project_mru` file). For each entry with a custom name, a tab character is appended as a separator, followed by the custom name. These files live in `~/.local/share/rstudio/monitored/lists/` for Mac/Linux and `%localappdata%\RStudio\monitored\lists\` on Windows.

If there isn't a `project_name_mru` file at startup it will be seeded with the contents of the old `project_mru` file. Both files will be kept up-to-date as projects are opened and closed (this allows downgrading RStudio without seeing an out-of-date project mru). If the recent projects list is cleared by the user, both files will be emptied.

Here's the options dialog with no custom name entered, followed by an example with a name:

<img width="578" alt="screenshot of project options dialog with no custom name set" src="https://github.com/rstudio/rstudio/assets/10569626/49812ecd-bd1e-4531-80b6-033ec8df0f3d">

<img width="578" alt="screenshot of project options dialog with a custom name" src="https://github.com/rstudio/rstudio/assets/10569626/65dec130-aec3-4171-abf9-e6c4465573ac">

The name updates in all places when the project options dialog is closed (no need to restart). This PR includes minor fixes for satellite editor windows; they weren't updating the title when the git branch changed and didn't show the full title when a session with existing satellite editor windows was started up (#14029).

The title display logic (e.g. for the main window, satellite editor windows, the dock icon, and browser tab for server) is the same as before, except that when a custom name is available for a project, it will be shown where the folder name was previously being shown. Here's an example with Mac Desktop:

<img width="767" alt="mac-title" src="https://github.com/rstudio/rstudio/assets/10569626/029fc5f7-da11-4a7f-aa8b-432ccf71bdc6">

Here's the project dropdown button/menu showing a custom name for the current project. Note that the differences with more or less path info being shown is existing behavior, I didn't change that. I opted to append the custom name to the MRUs rather than replacing the folder name info (showing both here should provide maximum context for choosing a file, I think).

<img width="276" alt="dropdown" src="https://github.com/rstudio/rstudio/assets/10569626/bbabfac5-9e80-4c20-84b1-2f6c9961aa95">

### Automated Tests

None yet; I did create unit tests for the core MRU code (`MruList.cpp`) that I modified before I changed it, then added more tests for the new functionality (which is to support adding extra data after a separator character; the extra data doesn't enter into the uniqueness check for each entry, and can also be updated without moving the entry's position in the list).

### QA Notes

See the spec for more info (search for "IDE Spec: Customize R Project Names" in Confluence).

In general, if a custom name isn't entered, there should be no differences in behaviour (except for the bug fixes related to window titles not always updating when current git branch is changed).

This PR doesn't touch the pro-specific places that show project names: the Workbench Home page and the Project Sharing UI, so those continue showing folder-names for projects; see https://github.com/rstudio/rstudio-pro/issues/5589

### Documentation

TBD

### Checklist
